Meta/generate_encoding_indices.py

```diff
@@ -0,0 +1,309 @@
+
+    content += "}\n"
+
+    with open(output_path, "w") as f:
+        f.write(content)
+
+
+def generate_table_accessor(name: str, table: LookupTable) -> str:
+    size = len(table.code_points)
+
+    if table.first_pointer > 0:
+        return f"""
+Optional<u32> index_{name}_code_point(u32 pointer)
+{{
+    if (pointer < s_{name}_index_first_pointer || pointer - s_{name}_index_first_pointer >= s_{name}_index.size())
+        return {{}};
+    auto value = s_{name}_index[pointer - s_{name}_index_first_pointer];
+    if (value == 0xfffd)
+        return {{}};
+    return value;
+}}
+"""
+    else:
+        return f"""
+Optional<u32> index_{name}_code_point(u32 pointer)
+{{
+    if (pointer >= s_{name}_index.size())
+        return {{}};
+    auto value = s_{name}_index[pointer];
+    if (value == 0xfffd)
+        return {{}};
+    return value;
+}}
+"""
+
+
+def generate_inverse_table_accessor(name: str, table: LookupTable) -> str:
+    if table.first_pointer > 0:
+        return f"""
+Optional<u32> code_point_{name}_index(u32 code_point)
+{{
+    for (u32 i = 0; i < s_{name}_index.size(); ++i) {{
+        if (s_{name}_index[i] == code_point) {{
+            return s_{name}_index_first_pointer + i;
+        }}
+    }}
+    return {{}};
+}}
+"""
+    else:
+        return f"""
+Optional<u32> code_point_{name}_index(u32 code_point)
+{{
+    for (u32 i = 0; i < s_{name}_index.size(); ++i) {{
+        if (s_{name}_index[i] == code_point) {{
+            return i;
+        }}
+    }}
+    return {{}};
+}}
+"""
+
+
+def generate_implementation_file(tables: LookupTables, output_path: Path) -> None:
+    content = """
+#include <LibTextCodec/LookupTables.h>
+
+namespace TextCodec {
+"""
+
+    for name, table in tables.indexes.items():
+        if table.generate_accessor:
+            content += generate_table_accessor(name, table)
+        if table.generate_inverse_accessor:
+            content += generate_inverse_table_accessor(name, table)
+
+    content += "\n}\n"
+
+    with open(output_path, "w") as f:
+        f.write(content)
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Generate text codec lookup tables")
+    parser.add_argument(
+        "--generated-header-path", required=True, help="Path to the lookup table header file to generate"
+    )
+    parser.add_argument(
+        "--generated-implementation-path",
+        required=True,
+        help="Path to the lookup table implementation file to generate",
+    )
+    parser.add_argument("--json-path", required=True, help="Path to the JSON file to read from")
+
+    args = parser.parse_args()
+
+    with open(args.json_path, "r") as f:
+        data = json.load(f)
+
+    gb18030_table = prepare_table(data["gb18030"], GenerateAccessor.YES)
+
+    # FIXME: Encoding specification is not updated to GB-18030-2022 yet (https://github.com/whatwg/encoding/issues/312)
+    # NOTE: See https://commits.webkit.org/264918@main
+    gb18030_updates = {
```

Comment:
fwiw the encoding spec has been updated to reflect this Unicode recommendation. <https://github.com/whatwg/encoding/commit/2c3853e461afd718be198772170d024e427aee21>